Fix test compilation and lint failures with Rust 1.65.0

## Motivation

Zebra tests are failing with the Rust 1.65.0 release due to compilation errors:

https://github.com/ZcashFoundation/zebra/actions/runs/3387720824/jobs/5628827965#step:15:17284
<details>

> error: linking with `cc` failed: exit status: 1
> Error:   |
>   = note: "cc" "-m64" "/tmp/rustcBSeDle/symbols.o" "/tmp/rustdoctest3oeRg5/rust_out.rust_out.0b343d98-cgu.0.rcgu.o" "/tmp/rustdoctest3oeRg5/rust_out.471l5hwfdd61rvgf.rcgu.o" "-Wl,--as-needed" "-L" "/home/runner/work/zebra/zebra/target/debug/deps" "-L" "/home/runner/work/zebra/zebra/target/debug/deps" "-L" "/home/runner/work/zebra/zebra/target/debug/build/cxx-79a04422bdc932ed/out" "-L" "/home/runner/work/zebra/zebra/target/debug/build/libgit2-sys-dc2d8e534aa10880/out/build" "-L" "/home/runner/work/zebra/zebra/target/debug/build/librocksdb-sys-bc30ccf58b88f489/out" "-L" "/home/runner/work/zebra/zebra/target/debug/build/link-cplusplus-11343166a431a999/out" "-L" "/home/runner/work/zebra/zebra/target/debug/build/ring-1efd227b1ba0ef85/out" "-L" "/home/runner/work/zebra/zebra/target/debug/build/secp256k1-sys-fbf0283a6d94ce6a/out" "-L" "/home/runner/work/zebra/zebra/target/debug/build/zcash_script-7ffafdef64afbb09/out" "-L" 
> ....
> "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-01209bdfe9867748.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-5dba08c0e996feed.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-fcd555fdecc4f69e.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-63179a22dd24cae2.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-1ced5827449ff915.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-430d8655a2cd2488.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-e60518d2760eb01e.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-1ad943f92a4b1081.rlib" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-cce2c773caa14f2f.rlib" "-Wl,-Bdynamic" "-lstdc++" "-lstdc++" "-lstdc++" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/home/runner/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/rustdoctest3oeRg5/rust_out" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs"
>   = note: collect2: error: ld returned 1 exit status

</details>

There also seem to be issues related to changes in the Clippy lint:

```
error: the Err-variant returned from this function is very large
--> zebrad/src/components/sync.rs:1027:10
|
1027 | ) -> Result<(), BlockDownloadVerifyError> {
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the Err-variant is at least 144 bytes
```

https://github.com/ZcashFoundation/zebra/actions/runs/3387578258/jobs/5628508335#step:8:501

--- a/.github/workflows/continous-integration-os.patch.yml
+++ b/.github/workflows/continous-integration-os.patch.yml
@@ -15,16 +15,18 @@ on:
 
 jobs:
   test:
-    name: Test ${{ matrix.rust }} on ${{ matrix.os }}
-    runs-on: ${{ matrix.os }}
+    name: Test ${{ matrix.rust }} on ${{ matrix.os }}${{ matrix.features }}
     strategy:
       matrix:
         # TODO: Windows was removed for now, see https://github.com/ZcashFoundation/zebra/issues/3801
         os: [ubuntu-latest, macos-latest]
         rust: [stable, beta]
+        features: ["", " --features getblocktemplate-rpcs"]
         exclude:
           - os: macos-latest
             rust: beta
+          - os: macos-latest
+            features: " --features getblocktemplate-rpcs"
 
     steps:
       - run: 'echo "No build required"'

--- a/.github/workflows/continous-integration-os.yml
+++ b/.github/workflows/continous-integration-os.yml
@@ -53,7 +53,7 @@ env:
 
 jobs:
   test:
-    name: Test ${{ matrix.rust }} on ${{ matrix.os }}
+    name: Test ${{ matrix.rust }} on ${{ matrix.os }}${{ matrix.features }}
     # The large timeout is to accommodate:
     # - Windows builds (75 minutes, typically 30-50 minutes), and
     # - parameter downloads (an extra 90 minutes, but only when the cache expires)
@@ -65,6 +65,7 @@ jobs:
         # TODO: Windows was removed for now, see https://github.com/ZcashFoundation/zebra/issues/3801
         os: [ubuntu-latest, macos-latest]
         rust: [stable, beta]
+        features: ["", " --features getblocktemplate-rpcs"]
         exclude:
         # We're excluding macOS for the following reasons:
         # - the concurrent macOS runner limit is much lower than the Linux limit
@@ -72,6 +73,8 @@ jobs:
         # - macOS is a second-tier Zebra support platform
           - os: macos-latest
             rust: beta
+          - os: macos-latest
+            features: " --features getblocktemplate-rpcs"
 
     steps:
       - uses: actions/checkout@v3.1.0
@@ -139,9 +142,11 @@ jobs:
       # Modified from:
       # https://github.com/zcash/librustzcash/blob/c48bb4def2e122289843ddb3cb2984c325c03ca0/.github/workflows/ci.yml#L20-L33
       #
-      # TODO: split get-params-path and download-params examples into their own crate,
-      #       to speed up compilation
-      #       compile examples in release mode, to speed up downloads
+      # TODO:
+      # - split Fetch/Cache Zcash parameters into their own job,
+      #   and use `concurrency:` to limit it to one job per OS
+      # - split get-params-path and download-params examples into their own crate,
+      #   to speed up compilation
       - name: Fetch path to Zcash parameters
         working-directory: ./zebra-consensus
         shell: bash
@@ -157,28 +162,24 @@ jobs:
         working-directory: ./zebra-consensus
         run: cargo run --example download-params
 
-      - name: Run tests
+      - name: Run tests${{ matrix.features }}
         uses: actions-rs/cargo@v1.0.3
         with:
           command: test
-          args: --verbose --workspace -- --nocapture
-
-      - name: Run tests with getblocktemplate-rpcs feature
-        uses: actions-rs/cargo@v1.0.3
-        with:
-          command: test
-          args: --features getblocktemplate-rpcs --verbose --workspace -- --nocapture
+          args: ${{ matrix.features }} --verbose --workspace -- --nocapture
 
       # Explicitly run any tests that are usually #[ignored]
 
-      - name: Run zebrad large sync tests
+      - name: Run zebrad large sync tests${{ matrix.features }}
         # Skip the entire step on Ubuntu and Windows, because the test would be skipped anyway due to ZEBRA_SKIP_NETWORK_TESTS
+        # Currently, this also skips large sync with `getblocktemplate-rpcs`,
+        # but that is already covered by the Docker tests.
         if: matrix.os == 'macos-latest'
         uses: actions-rs/cargo@v1.0.3
         with:
           command: test
           # Note: this only runs the zebrad acceptance tests, because re-running all the test binaries is slow on Windows
-          args: --verbose --package zebrad --test acceptance -- --nocapture --include-ignored sync_large_checkpoints_
+          args: ${{ matrix.features }} --verbose --package zebrad --test acceptance -- --nocapture --include-ignored sync_large_checkpoints_
 
   # Install Zebra with lockfile dependencies, with no caching and default features
   install-from-lockfile-no-cache:

--- a/zebra-test/src/mock_service.rs
+++ b/zebra-test/src/mock_service.rs
@@ -76,7 +76,9 @@ const DEFAULT_PROXY_CHANNEL_SIZE: usize = 100;
 ///
 /// Note that if a test checks that no requests are received, each check has to wait for this
 /// amount of time, so this may affect the test execution time.
-pub const DEFAULT_MAX_REQUEST_DELAY: Duration = Duration::from_millis(25);
+///
+/// We've seen delays up to 67ms on busy Linux and macOS machines.
+pub const DEFAULT_MAX_REQUEST_DELAY: Duration = Duration::from_millis(150);
 
 /// An internal type representing the item that's sent in the [`broadcast`] channel.
 ///

--- a/zebrad/src/components/sync.rs
+++ b/zebrad/src/components/sync.rs
@@ -990,6 +990,7 @@ where
     /// Handles a response for a requested block.
     ///
     /// See [`Self::handle_response`] for more details.
+    #[allow(unknown_lints)]
     #[allow(clippy::result_large_err)]
     fn handle_block_response(
         &mut self,
@@ -1008,6 +1009,7 @@ where
     /// Handles a response to block hash submission, passing through any extra hashes.
     ///
     /// See [`Self::handle_response`] for more details.
+    #[allow(unknown_lints)]
     #[allow(clippy::result_large_err)]
     fn handle_hash_response(
         response: Result<IndexSet<block::Hash>, BlockDownloadVerifyError>,
@@ -1024,6 +1026,7 @@ where
     /// so that the synchronization can continue normally.
     ///
     /// Returns `Err` if an unexpected error occurred, to force the synchronizer to restart.
+    #[allow(unknown_lints)]
     #[allow(clippy::result_large_err)]
     fn handle_response<T>(
         response: Result<T, BlockDownloadVerifyError>,

--- a/zebrad/src/components/sync.rs
+++ b/zebrad/src/components/sync.rs
@@ -990,6 +990,7 @@ where
     /// Handles a response for a requested block.
     ///
     /// See [`Self::handle_response`] for more details.
+    #[allow(clippy::result_large_err)]
     fn handle_block_response(
         &mut self,
         response: Result<(Height, block::Hash), BlockDownloadVerifyError>,
@@ -1007,6 +1008,7 @@ where
     /// Handles a response to block hash submission, passing through any extra hashes.
     ///
     /// See [`Self::handle_response`] for more details.
+    #[allow(clippy::result_large_err)]
     fn handle_hash_response(
         response: Result<IndexSet<block::Hash>, BlockDownloadVerifyError>,
     ) -> Result<IndexSet<block::Hash>, BlockDownloadVerifyError> {
@@ -1022,6 +1024,7 @@ where
     /// so that the synchronization can continue normally.
     ///
     /// Returns `Err` if an unexpected error occurred, to force the synchronizer to restart.
+    #[allow(clippy::result_large_err)]
     fn handle_response<T>(
         response: Result<T, BlockDownloadVerifyError>,
     ) -> Result<(), BlockDownloadVerifyError> {

